source/common/http/conn_pool_grid.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "common/http/conn_pool_base.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Http {
+
+// An HTTP connection pool which will handle the connectivity grid of
+// [WiFi / cellular] [ipv4 / ipv6] [QUIC / TCP].
+// Currently only [QUIC / TCP are handled]
+class ConnectivityGrid : public ConnectionPool::Instance {
+public:
+  struct ConnectivityOptions {
+    explicit ConnectivityOptions(const std::vector<Http::Protocol>& protocols)
+        : protocols_(protocols) {}
+    std::vector<Http::Protocol> protocols_;
+  };
+
+  using PoolIterator = std::list<ConnectionPool::InstancePtr>::iterator;
+
+  // This is a class which wraps a caller's connection pool callbacks to
+  // auto-retry pools in the case of connection failure.
+  //
+  // It also relays cancellation calls between the original caller and the
+  // pools currently trying to connect.
+  //
+  // TODO(#15649) currently this tries one connection at a time.
+  // It should instead have a timer of its own and start the second connection
+  // in parallel after a suitable delay.
+  class WrapperCallbacks : public ConnectionPool::Callbacks, public ConnectionPool::Cancellable {
+  public:
+    WrapperCallbacks(ConnectivityGrid& grid, Http::ResponseDecoder& decoder, PoolIterator pool_it,
+                     ConnectionPool::Callbacks& callbacks);
+
+    // ConnectionPool::Callbacks
+    void onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                       absl::string_view transport_failure_reason,
+                       Upstream::HostDescriptionConstSharedPtr host) override;
+    void onPoolReady(RequestEncoder& encoder, Upstream::HostDescriptionConstSharedPtr host,
+                     const StreamInfo::StreamInfo& info,
+                     absl::optional<Http::Protocol> protocol) override;
+
+    // ConnectionPool::Cancellable
+    void cancel(Envoy::ConnectionPool::CancelPolicy cancel_policy) override;
+
+    ConnectionPool::Instance& pool() { return **pool_it_; }
+    // The owning grid.
+    ConnectivityGrid& grid_;
+    // The decoder for the original newStream, needed to create streams on subsequent pools.
+    Http::ResponseDecoder& decoder_;
+    // The callbacks from the original caller, which must get onPoolFailure or
+    // onPoolReady unless there is call to cancel().
+    ConnectionPool::Callbacks& inner_callbacks_;
+    // The current pool being connected to.
+    PoolIterator pool_it_;
+    // The handle to cancel the request to the current pool.
+    // This is owned by the pool which created it.
+    Cancellable* cancellable_;
```

Comment:
Since this can't be null, should it be a reference instead (apologies if this is wrong, it's a part of the Envoy style guide that I'm not used to)

Reply:
Yeah, Envoy is generally pretty anti-pointer, but this is set by one of the rare Envoy functions which returns a pointer, and we change what it points to, so I'm inclined to leave it as is. swapping the thing a reference refers to just feels wrong.  happy to be overriden by Matt here.

Reply:
Thanks for the explanation, happy to defer to the two of you here :)

Reply:
Optimally this would change to an RAII thing, but this is fine!